internal/operator-controller/authorization/rbac.go

```diff
@@ -0,0 +1,576 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	apimachyaml "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	rbacinternal "k8s.io/kubernetes/pkg/apis/rbac"
+	rbacv1helpers "k8s.io/kubernetes/pkg/apis/rbac/v1"
+	rbacregistry "k8s.io/kubernetes/pkg/registry/rbac"
+	"k8s.io/kubernetes/pkg/registry/rbac/validation"
+	rbac "k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	ocv1 "github.com/operator-framework/operator-controller/api/v1"
+)
+
+type PreAuthorizer interface {
+	PreAuthorize(ctx context.Context, ext *ocv1.ClusterExtension, manifestReader io.Reader) ([]ScopedPolicyRules, error)
+}
+
+type ScopedPolicyRules struct {
+	Namespace    string
+	MissingRules []rbacv1.PolicyRule
+}
+
+var objectVerbs = []string{"get", "patch", "update", "delete"}
+
+type rbacPreAuthorizer struct {
+	authorizer   authorizer.Authorizer
+	ruleResolver validation.AuthorizationRuleResolver
+	restMapper   meta.RESTMapper
+}
+
+func NewRBACPreAuthorizer(cl client.Client) PreAuthorizer {
+	return &rbacPreAuthorizer{
+		authorizer:   newRBACAuthorizer(cl),
+		ruleResolver: newRBACRulesResolver(cl),
+		restMapper:   cl.RESTMapper(),
+	}
+}
+
+// PreAuthorize validates whether the current user/request satisfies the necessary permissions
+// as defined by the RBAC policy. It examines the user’s roles, resource identifiers, and
+// the intended action to determine if the operation is allowed.
+//
+// Return Value:
+//   - nil: indicates that the authorization check passed and the operation is permitted.
+//   - non-nil error: indicates that an error occurred during the permission evaluation process
+//     (for example, a failure decoding the manifest or other internal issues). If the evaluation
+//     completes successfully but identifies missing rules, then a nil error is returned along with
+//     the list (or slice) of missing rules. Note that in some cases the error may encapsulate multiple
+//     evaluation failures
+func (a *rbacPreAuthorizer) PreAuthorize(ctx context.Context, ext *ocv1.ClusterExtension, manifestReader io.Reader) ([]ScopedPolicyRules, error) {
+	var allMissingPolicyRules = []ScopedPolicyRules{}
+	dm, err := a.decodeManifest(manifestReader)
+	if err != nil {
+		return nil, err
+	}
+	manifestManager := &user.DefaultInfo{Name: fmt.Sprintf("system:serviceaccount:%s:%s", ext.Spec.Namespace, ext.Spec.ServiceAccount.Name)}
+	attributesRecords := dm.asAuthorizationAttributesRecordsForUser(manifestManager, ext)
+
+	var preAuthEvaluationErrors []error
+	missingRules, err := a.authorizeAttributesRecords(ctx, attributesRecords)
+	if err != nil {
+		preAuthEvaluationErrors = append(preAuthEvaluationErrors, err)
+	}
+
+	ec := escalationChecker{
+		authorizer:        a.authorizer,
+		ruleResolver:      a.ruleResolver,
+		extraClusterRoles: dm.clusterRoles,
+		extraRoles:        dm.roles,
+	}
```

Comment:
suggestion: I'd put this behind a private method, e.g. 

```
func (a *rbacPreAuthorizer) escalationCheckerFor(dm *decodedManifest) *escalationChecker
```

Reply:
done in 352de66a